arangod/Metrics/MetricsFeature.cpp

```diff
@@ -0,0 +1,524 @@
+    nameVersionTable.try_emplace(
+        "arangodb_server_statistics_server_uptime_total",
+        "arangodb_server_statistics_server_uptime");
+    // And one for rocksdb:
+    nameVersionTable.try_emplace("rocksdb_engine_throttle_bps",
+                                 "rocksdbengine_throttle_bps");
+
+    v2suppressions = {
+        // Note that if we ever need to suppress a metric which is
+        // coming from the statistics feature further code is needed
+        // there. This list is only considered for the normally registered
+        // metrics in the MetricsFeature.
+        "arangodb_maintenance_phase1_accum_runtime_msec_total",
+        "arangodb_maintenance_phase2_accum_runtime_msec_total",
+        "arangodb_maintenance_agency_sync_accum_runtime_msec_total",
+        "arangodb_maintenance_action_accum_runtime_msec_total",
+        "arangodb_maintenance_action_accum_queue_time_msec_total",
+        "arangodb_agency_supervision_accum_runtime_msec_total",
+        "arangodb_agency_supervision_accum_runtime_wait_for_replication_msec_"
+        "total",
+        "arangodb_load_current_accum_runtime_msec_total",
+        "arangodb_load_plan_accum_runtime_msec_total",
+        "arangodb_aql_slow_query_total",
+        "arangodb_scheduler_jobs_dequeued",
+        "arangodb_scheduler_jobs_submitted",
+        "arangodb_scheduler_jobs_done",
+    };
+
+    v1suppressions = {
+        "arangodb_scheduler_jobs_dequeued_total",
+        "arangodb_scheduler_jobs_submitted_total",
+        "arangodb_scheduler_jobs_done_total",
+    };
+  } catch (std::exception const& e) {
+    LOG_TOPIC("efd51", ERR, Logger::MEMORY)
+        << "failed to allocate and populate the metrics v1/v2 translation "
+           "table "
+        << e.what();
+    FATAL_ERROR_EXIT();
+  }
+}
+
+void MetricsFeature::collectOptions(std::shared_ptr<options::ProgramOptions> options) {
+  _serverStatistics =
+      std::make_unique<ServerStatistics>(*this, StatisticsFeature::time());
+
+  options
+      ->addOption("--server.export-metrics-api", "turn metrics API on or off",
+                  new options::BooleanParameter(&_export),
+                  arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden))
+      .setIntroducedIn(30600);
+
+  options
+      ->addOption("--server.export-read-write-metrics",
+                  "turn metrics for document read/write metrics on or off",
+                  new options::BooleanParameter(&_exportReadWriteMetrics),
+                  arangodb::options::makeDefaultFlags(arangodb::options::Flags::Hidden))
+      .setIntroducedIn(30707);
+}
+
+std::shared_ptr<Metric> MetricsFeature::doAdd(Builder& builder) {
+  auto metric = builder.build();
+  MetricKey key{metric->name(), metric->labels()};
+  bool success = false;
+  {
+    std::lock_guard guard(_lock);
+    success = _registry.try_emplace(key, metric).second;
+  }
+  if (!success) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, builder.type().data() +
```

Comment:
I don't think this is safe.
From https://en.cppreference.com/w/cpp/string/basic_string_view/data:
> Unlike std::basic_string::data() and string literals, data() may return a pointer to a buffer that is not null-terminated. Therefore it is typically a mistake to pass data() to a routine that takes just a const CharT* and expects a null-terminated string.

Reply:
But we know our `type()` string, its string literal, so everything fine

Reply:
Ok, it may work for now, but it is not necessarily future proof.
Can't we explicitly convert it into an `std::string(builder.type().data(), builder.type().size())` here? That will enable it to work with arbitrary types that will be added in the future.

Reply:
call string ctor now